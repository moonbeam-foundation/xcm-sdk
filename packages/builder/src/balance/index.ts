export * from './BalanceBuilder';
export * from './Erc20Abi';
export * from './BalanceBuilder.interfaces';
