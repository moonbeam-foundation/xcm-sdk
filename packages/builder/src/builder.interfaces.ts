import type {
  AnyParachain,
  AssetAmount,
  Chain,
} from '@moonbeam-network/xcm-types';
import type { ApiPromise } from '@polkadot/api';

export interface ConfigBuilder<Config, Params = BuilderParams> {
  build: (params: Params) => Config;
}

export interface BuilderParams<IChain extends Chain = AnyParachain> {
  asset: AssetAmount;
  destination: IChain;
  destinationAddress: string;
  destinationApi?: ApiPromise;
  fee: AssetAmount;
  source: IChain;
  sourceAddress: string;
  sourceApi?: ApiPromise;
}
