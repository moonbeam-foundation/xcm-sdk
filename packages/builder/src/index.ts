import '@polkadot/api-augment';

export * from './asset-min';
export * from './balance';
export * from './builder.interfaces';
export * from './contract';
export * from './extrinsic';
export * from './fee';
export * from './mrl';
export * from './types';
