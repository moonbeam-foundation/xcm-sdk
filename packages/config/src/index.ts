export * from './ConfigService';
export * from './assets';
export * from './chains';
export * from './config.utils';
export * from './mrl-configs';
export * from './types/AssetRoute';
export * from './types/MrlAssetRoute';
export * from './types/ChainRoutes';
export * from './types/MrlChainRoutes';
export * from './xcm-configs';
