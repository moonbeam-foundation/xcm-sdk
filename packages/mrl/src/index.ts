export * from './mrl';
export * from './mrl.interfaces';
