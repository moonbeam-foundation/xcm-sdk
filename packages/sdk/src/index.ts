export * from './getTransferData/getSourceData';
export * from './getTransferData/getDestinationData';
export * from './getTransferData/getTransferData.utils';
export * from './sdk';
export * from './sdk.interfaces';
export * from './services/polkadot';
export * from './services/evm';
