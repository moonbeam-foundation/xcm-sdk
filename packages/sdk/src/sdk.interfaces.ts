import type { AnyChain, AssetAmount } from '@moonbeam-network/xcm-types';
import type { Signer } from '@polkadot/api/types';
import type { IKeyringPair } from '@polkadot/types/types';
import type { WalletClient } from 'viem';

export type EvmSigner = WalletClient;

export interface Signers {
  evmSigner?: EvmSigner;
  polkadotSigner?: Signer | IKeyringPair;
}

export interface TransferData {
  destination: DestinationChainTransferData;
  getEstimate(amount: number | string): AssetAmount;
  max: AssetAmount;
  min: AssetAmount;
  source: SourceChainTransferData;
  transfer(amount: bigint | number | string, signers: Signers): Promise<string>;
}

export interface SourceChainTransferData extends ChainTransferData {
  destinationFee: AssetAmount;
  destinationFeeBalance: AssetAmount;
  feeBalance: AssetAmount;
  max: AssetAmount;
}

export interface SovereignAccountBalance {
  feeAssetBalance: bigint | undefined;
  transferAssetBalance: bigint;
}

export interface DestinationChainTransferData extends ChainTransferData {
  sovereignAccountBalances?: SovereignAccountBalance;
}

export interface ChainTransferData {
  balance: AssetAmount;
  chain: AnyChain;
  existentialDeposit?: AssetAmount;
  fee: AssetAmount;
  min: AssetAmount;
}
