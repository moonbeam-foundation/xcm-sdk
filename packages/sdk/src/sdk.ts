import { ConfigService, xcmRoutesMap } from '@moonbeam-network/xcm-config';
import {
  type AnyAsset,
  type AnyChain,
  type AnyParachain,
  type AssetAmount,
  type Ecosystem,
  EvmParachain,
} from '@moonbeam-network/xcm-types';
import { getAssetsBalances } from './getTransferData/getSourceData';
import { getTransferData } from './getTransferData/getTransferData';
import type { TransferData } from './sdk.interfaces';

const DEFAULT_SERVICE = new ConfigService({ routes: xcmRoutesMap });

export interface SdkOptions {
  configService?: ConfigService;
  ecosystem?: Ecosystem;
}

export function Sdk({ configService, ecosystem }: SdkOptions = {}) {
  const service = configService ?? DEFAULT_SERVICE;
  const assets = service.getEcosystemAssets(ecosystem);

  return {
    assets,
    setAsset(asset: string | AnyAsset) {
      const sources = service.getSourceChains({ asset, ecosystem });

      return {
        sources,
        setSource(source: string | AnyChain) {
          const destinations = service.getDestinationChains({
            asset,
            source,
          });

          return {
            destinations,
            setDestination(destination: string | AnyChain) {
              const route = service.getAssetRoute({
                asset,
                source,
                destination,
              });

              return {
                setAddresses({
                  sourceAddress,
                  destinationAddress,
                }: {
                  sourceAddress: string;
                  destinationAddress: string;
                }): Promise<TransferData> {
                  const sourceChain = service.getChain(source);

                  if (!EvmParachain.isAnyParachain(sourceChain)) {
                    throw new Error(
                      'Source chain should be a Parachain or EvmParachain',
                    );
                  }

                  if (!EvmParachain.isAnyParachain(route.destination.chain)) {
                    throw new Error(
                      'Destination chain should be a Parachain or EvmParachain',
                    );
                  }

                  return getTransferData({
                    route,
                    sourceAddress,
                    destinationAddress,
                  });
                },
              };
            },
          };
        },
      };
    },
  };
}

export async function getParachainBalances(
  chain: AnyParachain,
  address: string,
  service: ConfigService = DEFAULT_SERVICE,
): Promise<AssetAmount[]> {
  const routes = service.getChainRoutes(chain).getRoutes();
  const balances = await getAssetsBalances({
    chain,
    routes,
    address,
  });

  return balances;
}
