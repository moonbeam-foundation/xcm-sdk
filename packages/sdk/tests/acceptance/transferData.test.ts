import { describe, expect, it } from 'vitest';

import {
  alphanetRelay,
  laos,
  laosMainnet,
  moonbaseAlpha,
  moonbeam,
  unit,
} from '@moonbeam-network/xcm-config';
import type { AnyParachain, Asset } from '@moonbeam-network/xcm-types';
import { Sdk, type TransferData } from '../../src';
import {
  laosMainnetAddress,
  moonEvmAddress,
  substrateAddress,
} from './constants';

const transferDateTestConfig: {
  asset: Asset;
  source: AnyParachain;
  destination: AnyParachain;
  sourceAddress: string;
  destinationAddress: string;
}[] = [
  {
    asset: unit,
    source: alphanetRelay,
    sourceAddress: substrateAddress,
    destination: moonbaseAlpha,
    destinationAddress: moonEvmAddress,
  },
  {
    asset: unit,
    source: moonbaseAlpha,
    sourceAddress: moonEvmAddress,
    destination: alphanetRelay,
    destinationAddress: substrateAddress,
  },
  {
    asset: laos,
    source: moonbeam,
    sourceAddress: moonEvmAddress,
    destination: laosMainnet,
    destinationAddress: laosMainnetAddress,
  },
];

describe('sdk/transferData', () => {
  describe.each(transferDateTestConfig)(
    'asset $asset.key from source $source.name to destination $destination.name',
    ({ asset, source, sourceAddress, destination, destinationAddress }) => {
      it('should get expected balances', async () => {
        const transferData: TransferData = await Sdk()
          .setAsset(asset)
          .setSource(source)
          .setDestination(destination)
          .setAddresses({
            sourceAddress: sourceAddress,
            destinationAddress: destinationAddress,
          });

        expect(transferData).toMatchSnapshot();
      });
    },
  );
});
