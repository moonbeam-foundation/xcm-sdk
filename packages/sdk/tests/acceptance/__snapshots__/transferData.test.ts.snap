// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sdk/transferData > asset 'hdx' from source 'Hydration Alphanet' to destination 'Moonbase Alpha' > should get expected balances 1`] = `
{
  "destination": {
    "balance": _AssetAmount {
      "address": "0xFFFfFfff345Dc44DDAE98Df024Eb494321E73FcC",
      "amount": 0n,
      "decimals": 12,
      "ids": {
        "id": "69606720909260275826784788104880799692",
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "chain": _EvmParachain {
      "assets": Map {
        "alan" => _ChainAsset {
          "address": "0x9133c5a22024118804089f1fB752b7B2ce2a6351",
          "decimals": 18,
          "ids": {
            "palletInstance": 48,
          },
          "key": "alan",
          "min": undefined,
          "originSymbol": "ALAN",
          "symbol": undefined,
        },
        "ampe" => _ChainAsset {
          "address": "0xfFfFfffF7fee8415e1c2AC3A15C48D3546B95e16",
          "decimals": 12,
          "ids": {
            "id": "170050401128744171791743427490841452054",
          },
          "key": "ampe",
          "min": undefined,
          "originSymbol": "AMPE",
          "symbol": undefined,
        },
        "atom" => _ChainAsset {
          "address": "0xffffffffb7cdb201c395c238350568f17cfbd3b5",
          "decimals": 6,
          "ids": {
            "id": "244316754493307480955066032215622931381",
          },
          "key": "atom",
          "min": undefined,
          "originSymbol": "ATOM",
          "symbol": undefined,
        },
        "dev" => _ChainAsset {
          "address": "0x0000000000000000000000000000000000000802",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "dev",
          "min": 10000000000000000n,
          "originSymbol": "DEV",
          "symbol": undefined,
        },
        "lit" => _ChainAsset {
          "address": "0xfffFFfFF31103d490325BB0a8E40eF62e2F614C0",
          "decimals": 12,
          "ids": {
            "id": "65216491554813189869575508812319036608",
          },
          "key": "lit",
          "min": undefined,
          "originSymbol": "LIT",
          "symbol": undefined,
        },
        "maos" => _ChainAsset {
          "address": "0xffffffff06a14960bddc8610d0e694d39a3c682e",
          "decimals": 18,
          "ids": {
            "palletInstance": 48,
          },
          "key": "maos",
          "min": undefined,
          "originSymbol": "MAOS",
          "symbol": undefined,
        },
        "otp" => _ChainAsset {
          "address": "0xFfffffFfB3229c8E7657eABEA704d5e75246e544",
          "decimals": 12,
          "ids": {
            "id": "238111524681612888331172110363070489924",
          },
          "key": "otp",
          "min": undefined,
          "originSymbol": "OTP",
          "symbol": undefined,
        },
        "paring" => _ChainAsset {
          "address": "0xFFFffFfF8283448b3cB519Ca4732F2ddDC6A6165",
          "decimals": 18,
          "ids": {
            "id": "173481220575862801646329923366065693029",
          },
          "key": "paring",
          "min": undefined,
          "originSymbol": "PARING",
          "symbol": undefined,
        },
        "pica" => _ChainAsset {
          "address": "0xFFFfFFFF10DD5Fd142163a40Ce0dae8c56e2801f",
          "decimals": 6,
          "ids": {
            "id": "22417088946346045371238623691600461855",
          },
          "key": "pica",
          "min": undefined,
          "originSymbol": "PICA",
          "symbol": undefined,
        },
        "tt1" => _ChainAsset {
          "address": "0xfFffFfFf75976211C786fe4d73d2477e222786Ac",
          "decimals": 18,
          "ids": {
            "id": "156305701417244550631956600137082963628",
          },
          "key": "tt1",
          "min": undefined,
          "originSymbol": "TT1",
          "symbol": undefined,
        },
        "tur" => _ChainAsset {
          "address": "0xfFffffFf6448d0746f2a66342B67ef9CAf89478E",
          "decimals": 10,
          "ids": {
            "id": "133300872918374599700079037156071917454",
          },
          "key": "tur",
          "min": undefined,
          "originSymbol": "TUR",
          "symbol": undefined,
        },
        "unit" => _ChainAsset {
          "address": "0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080",
          "decimals": 12,
          "ids": {
            "id": "42259045809535163221576417993425387648",
          },
          "key": "unit",
          "min": undefined,
          "originSymbol": "Unit",
          "symbol": undefined,
        },
        "ftmwh" => _ChainAsset {
          "address": "0x566c1cebc6A4AFa1C122E039C4BEBe77043148Ee",
          "decimals": 18,
          "ids": {
            "palletInstance": 48,
          },
          "key": "ftmwh",
          "min": undefined,
          "originSymbol": "FTM.wh",
          "symbol": undefined,
        },
        "hdx" => _ChainAsset {
          "address": "0xFFFfFfff345Dc44DDAE98Df024Eb494321E73FcC",
          "decimals": 12,
          "ids": {
            "id": "69606720909260275826784788104880799692",
          },
          "key": "hdx",
          "min": undefined,
          "originSymbol": "HDX",
          "symbol": undefined,
        },
        "usdcwh" => _ChainAsset {
          "address": "0xE5dE10C4b744bac6b783fAF8d9B9fDFF14Acc3c9",
          "decimals": 6,
          "ids": {
            "palletInstance": 48,
          },
          "key": "usdcwh",
          "min": undefined,
          "originSymbol": "USDC.Wh",
          "symbol": undefined,
        },
        "agng" => _ChainAsset {
          "address": "0xFFfFFFFF38794F8c53fC7325ea07463dc6965e20",
          "decimals": 18,
          "ids": {
            "id": "75066649112131892397889252914026143264",
          },
          "key": "agng",
          "min": undefined,
          "originSymbol": "AGNG",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "contracts": undefined,
      "ecosystem": "alphanet-relay",
      "explorer": "https://moonbase.moonscan.io",
      "genesisHash": "0x91bc6e169807aaa54802737e1c504b2577d4fafedd5a02c10293b1cd60e39527",
      "id": 1287,
      "isEvmSigner": true,
      "isRelay": false,
      "isTestChain": true,
      "key": "moonbase-alpha",
      "name": "Moonbase Alpha",
      "parachainId": 1000,
      "rpc": "https://rpc.api.moonbase.moonbeam.network",
      "ss58Format": 1287,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": {
        "name": "Moonbeam",
      },
      "ws": [
        "wss://wss.api.moonbase.moonbeam.network",
      ],
    },
    "existentialDeposit": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "dev",
      "min": 10000000000000000n,
      "originSymbol": "DEV",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": "0xFFFfFfff345Dc44DDAE98Df024Eb494321E73FcC",
      "amount": 423284151910n,
      "decimals": 12,
      "ids": {
        "id": "69606720909260275826784788104880799692",
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": "0xFFFfFfff345Dc44DDAE98Df024Eb494321E73FcC",
      "amount": 0n,
      "decimals": 12,
      "ids": {
        "id": "69606720909260275826784788104880799692",
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "sovereignAccountBalances": undefined,
  },
  "getEstimate": [Function],
  "max": _AssetAmount {
    "address": undefined,
    "amount": 1587499959812n,
    "decimals": 12,
    "ids": {
      "id": 0,
    },
    "key": "hdx",
    "min": undefined,
    "originSymbol": "HDX",
    "symbol": undefined,
  },
  "min": _AssetAmount {
    "address": "0xFFFfFfff345Dc44DDAE98Df024Eb494321E73FcC",
    "amount": 423284151910n,
    "decimals": 12,
    "ids": {
      "id": "69606720909260275826784788104880799692",
    },
    "key": "hdx",
    "min": undefined,
    "originSymbol": "HDX",
    "symbol": undefined,
  },
  "source": {
    "balance": _AssetAmount {
      "address": undefined,
      "amount": 3000000000000n,
      "decimals": 12,
      "ids": {
        "id": 0,
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "chain": _Parachain {
      "assets": Map {
        "hdx" => _ChainAsset {
          "address": undefined,
          "decimals": 12,
          "ids": {
            "id": 0,
          },
          "key": "hdx",
          "min": undefined,
          "originSymbol": "HDX",
          "symbol": undefined,
        },
        "usdcwh" => _ChainAsset {
          "address": undefined,
          "decimals": 6,
          "ids": {
            "id": 1000001,
          },
          "key": "usdcwh",
          "min": undefined,
          "originSymbol": "USDC.Wh",
          "symbol": undefined,
        },
        "ftmwh" => _ChainAsset {
          "address": undefined,
          "decimals": 18,
          "ids": {
            "id": 1000002,
          },
          "key": "ftmwh",
          "min": undefined,
          "originSymbol": "FTM.wh",
          "symbol": undefined,
        },
        "usdtwh" => _ChainAsset {
          "address": undefined,
          "decimals": 6,
          "ids": {
            "id": 2,
          },
          "key": "usdtwh",
          "min": undefined,
          "originSymbol": "USDT.Wh",
          "symbol": undefined,
        },
        "dev" => _ChainAsset {
          "address": undefined,
          "decimals": 18,
          "ids": {
            "id": 1,
          },
          "key": "dev",
          "min": undefined,
          "originSymbol": "DEV",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "ecosystem": "alphanet-relay",
      "explorer": "https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fhydradx-moonbase-rpc.play.hydration.cloud#/explorer/query",
      "genesisHash": "0x025980095be141a99f983631c49271af15cab61c4ce0d73db73192443932669a",
      "isRelay": false,
      "isTestChain": true,
      "key": "hydration-Alphanet",
      "name": "Hydration Alphanet",
      "parachainId": 2034,
      "ss58Format": 63,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": undefined,
      "ws": [
        "wss://hydradx-moonbase-rpc.play.hydration.cloud",
      ],
    },
    "destinationFee": _AssetAmount {
      "address": undefined,
      "amount": 423284151910n,
      "decimals": 12,
      "ids": {
        "id": 0,
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "destinationFeeBalance": _AssetAmount {
      "address": undefined,
      "amount": 3000000000000n,
      "decimals": 12,
      "ids": {
        "id": 0,
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "existentialDeposit": _AssetAmount {
      "address": undefined,
      "amount": 1000000000000n,
      "decimals": 12,
      "ids": {
        "id": 0,
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": undefined,
      "amount": 412500040188n,
      "decimals": 12,
      "ids": {
        "id": 0,
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "feeBalance": _AssetAmount {
      "address": undefined,
      "amount": 3000000000000n,
      "decimals": 12,
      "ids": {
        "id": 0,
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "max": _AssetAmount {
      "address": undefined,
      "amount": 1587499959812n,
      "decimals": 12,
      "ids": {
        "id": 0,
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": undefined,
      "amount": 0n,
      "decimals": 12,
      "ids": {
        "id": 0,
      },
      "key": "hdx",
      "min": undefined,
      "originSymbol": "HDX",
      "symbol": undefined,
    },
  },
  "transfer": [Function],
}
`;

exports[`sdk/transferData > asset 'unit' from source 'Moonbase Alpha' to destination 'Alphanet Relay' > should get expected balances 1`] = `
{
  "destination": {
    "balance": _AssetAmount {
      "address": undefined,
      "amount": 0n,
      "decimals": 12,
      "ids": undefined,
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
    "chain": _Parachain {
      "assets": Map {
        "unit" => _ChainAsset {
          "address": undefined,
          "decimals": 12,
          "ids": undefined,
          "key": "unit",
          "min": undefined,
          "originSymbol": "Unit",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "ecosystem": "alphanet-relay",
      "explorer": undefined,
      "genesisHash": "0xe1ea3ab1d46ba8f4898b6b4b9c54ffc05282d299f89e84bd0fd08067758c9443",
      "isRelay": true,
      "isTestChain": true,
      "key": "alphanet-relay",
      "name": "Alphanet Relay",
      "parachainId": 0,
      "ss58Format": 42,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": undefined,
      "ws": [
        "wss://relay.api.moonbase.moonbeam.network",
        "wss://fro-moon-rpc-1-moonbase-relay-rpc-1.moonbase.ol-infra.network",
      ],
    },
    "existentialDeposit": _AssetAmount {
      "address": undefined,
      "amount": 10000000000n,
      "decimals": 12,
      "ids": undefined,
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": undefined,
      "amount": 50600000000n,
      "decimals": 12,
      "ids": undefined,
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": undefined,
      "amount": 0n,
      "decimals": 12,
      "ids": undefined,
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
    "sovereignAccountBalances": undefined,
  },
  "getEstimate": [Function],
  "max": _AssetAmount {
    "address": "0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080",
    "amount": 10000n,
    "decimals": 12,
    "ids": {
      "id": "42259045809535163221576417993425387648",
    },
    "key": "unit",
    "min": undefined,
    "originSymbol": "Unit",
    "symbol": undefined,
  },
  "min": _AssetAmount {
    "address": undefined,
    "amount": 60600000000n,
    "decimals": 12,
    "ids": undefined,
    "key": "unit",
    "min": undefined,
    "originSymbol": "Unit",
    "symbol": undefined,
  },
  "source": {
    "balance": _AssetAmount {
      "address": "0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080",
      "amount": 10000n,
      "decimals": 12,
      "ids": {
        "id": "42259045809535163221576417993425387648",
      },
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
    "chain": _EvmParachain {
      "assets": Map {
        "alan" => _ChainAsset {
          "address": "0x9133c5a22024118804089f1fB752b7B2ce2a6351",
          "decimals": 18,
          "ids": {
            "palletInstance": 48,
          },
          "key": "alan",
          "min": undefined,
          "originSymbol": "ALAN",
          "symbol": undefined,
        },
        "ampe" => _ChainAsset {
          "address": "0xfFfFfffF7fee8415e1c2AC3A15C48D3546B95e16",
          "decimals": 12,
          "ids": {
            "id": "170050401128744171791743427490841452054",
          },
          "key": "ampe",
          "min": undefined,
          "originSymbol": "AMPE",
          "symbol": undefined,
        },
        "atom" => _ChainAsset {
          "address": "0xffffffffb7cdb201c395c238350568f17cfbd3b5",
          "decimals": 6,
          "ids": {
            "id": "244316754493307480955066032215622931381",
          },
          "key": "atom",
          "min": undefined,
          "originSymbol": "ATOM",
          "symbol": undefined,
        },
        "dev" => _ChainAsset {
          "address": "0x0000000000000000000000000000000000000802",
          "decimals": 18,
          "ids": {
            "palletInstance": 3,
          },
          "key": "dev",
          "min": 10000000000000000n,
          "originSymbol": "DEV",
          "symbol": undefined,
        },
        "lit" => _ChainAsset {
          "address": "0xfffFFfFF31103d490325BB0a8E40eF62e2F614C0",
          "decimals": 12,
          "ids": {
            "id": "65216491554813189869575508812319036608",
          },
          "key": "lit",
          "min": undefined,
          "originSymbol": "LIT",
          "symbol": undefined,
        },
        "maos" => _ChainAsset {
          "address": "0xffffffff06a14960bddc8610d0e694d39a3c682e",
          "decimals": 18,
          "ids": {
            "palletInstance": 48,
          },
          "key": "maos",
          "min": undefined,
          "originSymbol": "MAOS",
          "symbol": undefined,
        },
        "otp" => _ChainAsset {
          "address": "0xFfffffFfB3229c8E7657eABEA704d5e75246e544",
          "decimals": 12,
          "ids": {
            "id": "238111524681612888331172110363070489924",
          },
          "key": "otp",
          "min": undefined,
          "originSymbol": "OTP",
          "symbol": undefined,
        },
        "paring" => _ChainAsset {
          "address": "0xFFFffFfF8283448b3cB519Ca4732F2ddDC6A6165",
          "decimals": 18,
          "ids": {
            "id": "173481220575862801646329923366065693029",
          },
          "key": "paring",
          "min": undefined,
          "originSymbol": "PARING",
          "symbol": undefined,
        },
        "pica" => _ChainAsset {
          "address": "0xFFFfFFFF10DD5Fd142163a40Ce0dae8c56e2801f",
          "decimals": 6,
          "ids": {
            "id": "22417088946346045371238623691600461855",
          },
          "key": "pica",
          "min": undefined,
          "originSymbol": "PICA",
          "symbol": undefined,
        },
        "tt1" => _ChainAsset {
          "address": "0xfFffFfFf75976211C786fe4d73d2477e222786Ac",
          "decimals": 18,
          "ids": {
            "id": "156305701417244550631956600137082963628",
          },
          "key": "tt1",
          "min": undefined,
          "originSymbol": "TT1",
          "symbol": undefined,
        },
        "tur" => _ChainAsset {
          "address": "0xfFffffFf6448d0746f2a66342B67ef9CAf89478E",
          "decimals": 10,
          "ids": {
            "id": "133300872918374599700079037156071917454",
          },
          "key": "tur",
          "min": undefined,
          "originSymbol": "TUR",
          "symbol": undefined,
        },
        "unit" => _ChainAsset {
          "address": "0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080",
          "decimals": 12,
          "ids": {
            "id": "42259045809535163221576417993425387648",
          },
          "key": "unit",
          "min": undefined,
          "originSymbol": "Unit",
          "symbol": undefined,
        },
        "ftmwh" => _ChainAsset {
          "address": "0x566c1cebc6A4AFa1C122E039C4BEBe77043148Ee",
          "decimals": 18,
          "ids": {
            "palletInstance": 48,
          },
          "key": "ftmwh",
          "min": undefined,
          "originSymbol": "FTM.wh",
          "symbol": undefined,
        },
        "hdx" => _ChainAsset {
          "address": "0xFFFfFfff345Dc44DDAE98Df024Eb494321E73FcC",
          "decimals": 12,
          "ids": {
            "id": "69606720909260275826784788104880799692",
          },
          "key": "hdx",
          "min": undefined,
          "originSymbol": "HDX",
          "symbol": undefined,
        },
        "usdcwh" => _ChainAsset {
          "address": "0xE5dE10C4b744bac6b783fAF8d9B9fDFF14Acc3c9",
          "decimals": 6,
          "ids": {
            "palletInstance": 48,
          },
          "key": "usdcwh",
          "min": undefined,
          "originSymbol": "USDC.Wh",
          "symbol": undefined,
        },
        "agng" => _ChainAsset {
          "address": "0xFFfFFFFF38794F8c53fC7325ea07463dc6965e20",
          "decimals": 18,
          "ids": {
            "id": "75066649112131892397889252914026143264",
          },
          "key": "agng",
          "min": undefined,
          "originSymbol": "AGNG",
          "symbol": undefined,
        },
      },
      "checkSovereignAccountBalances": false,
      "contracts": undefined,
      "ecosystem": "alphanet-relay",
      "explorer": "https://moonbase.moonscan.io",
      "genesisHash": "0x91bc6e169807aaa54802737e1c504b2577d4fafedd5a02c10293b1cd60e39527",
      "id": 1287,
      "isEvmSigner": true,
      "isRelay": false,
      "isTestChain": true,
      "key": "moonbase-alpha",
      "name": "Moonbase Alpha",
      "parachainId": 1000,
      "rpc": "https://rpc.api.moonbase.moonbeam.network",
      "ss58Format": 1287,
      "usesChainDecimals": false,
      "weight": undefined,
      "wh": {
        "name": "Moonbeam",
      },
      "ws": [
        "wss://wss.api.moonbase.moonbeam.network",
      ],
    },
    "destinationFee": _AssetAmount {
      "address": "0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080",
      "amount": 50600000000n,
      "decimals": 12,
      "ids": {
        "id": "42259045809535163221576417993425387648",
      },
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
    "destinationFeeBalance": _AssetAmount {
      "address": "0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080",
      "amount": 10000n,
      "decimals": 12,
      "ids": {
        "id": "42259045809535163221576417993425387648",
      },
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
    "existentialDeposit": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 0n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "dev",
      "min": 10000000000000000n,
      "originSymbol": "DEV",
      "symbol": undefined,
    },
    "fee": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 6922750000000n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "dev",
      "min": 10000000000000000n,
      "originSymbol": "DEV",
      "symbol": undefined,
    },
    "feeBalance": _AssetAmount {
      "address": "0x0000000000000000000000000000000000000802",
      "amount": 6543210000000000000n,
      "decimals": 18,
      "ids": {
        "palletInstance": 3,
      },
      "key": "dev",
      "min": 10000000000000000n,
      "originSymbol": "DEV",
      "symbol": undefined,
    },
    "max": _AssetAmount {
      "address": "0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080",
      "amount": 10000n,
      "decimals": 12,
      "ids": {
        "id": "42259045809535163221576417993425387648",
      },
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
    "min": _AssetAmount {
      "address": "0xFfFFfFff1FcaCBd218EDc0EbA20Fc2308C778080",
      "amount": 0n,
      "decimals": 12,
      "ids": {
        "id": "42259045809535163221576417993425387648",
      },
      "key": "unit",
      "min": undefined,
      "originSymbol": "Unit",
      "symbol": undefined,
    },
  },
  "transfer": [Function],
}
`;
